### PR TITLE
MEPTS-1410 | CXCA SCRN Changes for March 2024 Release

### PR DESCRIPTION
**Pull request by TinoMuianga**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 1]
```diff
@@ -77,6 +78,392 @@ private CohortDefinition getA() {
...
+    map.put("703", hivMetadata.getPositive().getConceptId());
+    map.put("2093", hivMetadata.getSuspectedCancerConcept().getConceptId());
+    map.put("165436", hivMetadata.getHumanPapillomavirusDna().getConceptId());
+
+    StringBuilder query =
+        new StringBuilder(
+            "SELECT final.patient_id "
+                + "FROM   ( "
+                + "           SELECT   via_or_hpv.patient_id, "
+                + "                    Max(via_or_hpv.screening_date) AS last_screening_date "
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java]: The requirements are not looking for the last screening date **(at least one)** 
Why using MAX function?

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 2]
```diff
@@ -77,6 +78,392 @@ private CohortDefinition getA() {
...
+    if (cxcascrnResult == CXCASCRNCohortQueries.CXCASCRNResult.NEGATIVE) {
+      query.append("    AND o.value_coded = ${664} ");
+    } else if (cxcascrnResult == CXCASCRNCohortQueries.CXCASCRNResult.POSITIVE) {
+      query.append("    AND o.value_coded = ${703} ");
+    } else if (cxcascrnResult == CXCASCRNCohortQueries.CXCASCRNResult.SUSPECTED) {
+      query.append("    AND o.value_coded = ${2093} ");
+    } else if (cxcascrnResult == CXCASCRNCohortQueries.CXCASCRNResult.ANY) {
+      query.append("    AND o.value_coded IN (${2093}, ${664}, ${703}) ");
+    }
+    query.append("                          AND        e.encounter_datetime  ");
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java:141]: If you are using DATE function on the select clause, it should be used on `e.encounter_datetime` verification's either.

Also check if there is a possibility of this date comes with time, like encounter 13/51. if not, there is no need to use this function

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 3]
```diff
@@ -77,6 +78,392 @@ private CohortDefinition getA() {
...
+            + "        AND            o.value_coded IN (${2093}, ${664}, ${703}) ) "
+            + "         OR          (  o2.concept_id = ${165436} "
+            + "        AND             o2.value_coded = ${664} ) )  "
+            + "        AND        e.location_id = :location "
+            + "        AND        e.encounter_datetime  ";
+    if (beforeStartDate) {
+      query += (" < :startDate ");
+    } else {
+      query += ("  BETWEEN :startDate AND  :endDate ");
+    }
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java]: Consider using ternary operator to check this boolean cases

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 4]
```diff
@@ -77,6 +78,392 @@ private CohortDefinition getA() {
...
+            + "                  AND e.encounter_type = ${28} "
+            + "                  AND o.concept_id IN ( ${2094}, ${165436} ) "
+            + "                  AND o.value_coded = ${664} "
+            + "                  AND e.location_id = :location "
+            + "                  AND e.encounter_datetime ";
+    if (beforeStartDate) {
+      query += (" < :startDate ");
+    } else {
+      query += ("  BETWEEN :startDate AND  :endDate ");
+    }
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java]: Ternary operator

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 5]
```diff
@@ -77,6 +78,392 @@ private CohortDefinition getA() {
...
+            + "        GROUP  BY patient_id) max_negative "
+            + "WHERE  NOT EXISTS (SELECT e.patient_id "
+            + "                   FROM   encounter e "
+            + "                              INNER JOIN obs o "
+            + "                                         ON o.encounter_id = e.encounter_id "
+            + "                   WHERE  e.voided = 0 "
+            + "                     AND o.voided = 0 "
+            + "                     AND e.encounter_type = ${28} "
+            + "                     AND o.concept_id = ${2094} "
+            + "                     AND o.value_coded IN ( ${703}, ${2093} ) "
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java:321]: Why you are not checking 165436 in this other possible results?

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 6]
```diff
@@ -77,6 +78,392 @@ private CohortDefinition getA() {
...
+    map.put("703", hivMetadata.getPositive().getConceptId());
+    map.put("2093", hivMetadata.getSuspectedCancerConcept().getConceptId());
+    map.put("165436", hivMetadata.getHumanPapillomavirusDna().getConceptId());
+
+    String query =
+        "SELECT patient_id "
+            + "FROM   (SELECT negative.patient_id, "
+            + "               Max(negative.last_negative_result_date) AS negative_result_Date "
+            + "        FROM   (SELECT p.patient_id, "
+            + "                       Max(e.encounter_datetime) AS last_negative_result_date "
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java]: Choose one of the clauses to pick the max date, I recommend to pick on the first one (outer)

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 7]
```diff
@@ -77,6 +78,392 @@ private CohortDefinition getA() {
...
+    map.put("703", hivMetadata.getPositive().getConceptId());
+    map.put("2093", hivMetadata.getSuspectedCancerConcept().getConceptId());
+    map.put("165436", hivMetadata.getHumanPapillomavirusDna().getConceptId());
+
+    String query =
+        "SELECT patient_id "
+            + "FROM   (SELECT suspected.patient_id, "
+            + "               Max(suspected.last_suspected_result_date) AS suspected_result_Date "
+            + "        FROM   (SELECT p.patient_id, "
+            + "                       Max(e.encounter_datetime) AS last_suspected_result_date "
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java]: Same here

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 8]
```diff
@@ -77,6 +78,392 @@ private CohortDefinition getA() {
...
+            + "                     AND o.value_coded IN ( ${703}, ${2093} ) "
+            + "                     AND e.location_id = :location "
+            + "                     AND max_negative.patient_id = e.patient_id "
+            + "                     AND e.encounter_datetime >= "
+            + "                         max_negative.negative_result_date "
+            + "                     AND e.encounter_datetime ";
+    if (beforeStartDate) {
+      query += (" < :startDate ) ");
+    } else {
+      query += ("  BETWEEN :startDate AND  :endDate )");
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java]: this verification is not needed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 9]
```diff
@@ -77,6 +78,392 @@ private CohortDefinition getA() {
...
+            + "                   WHERE  e.voided = 0 "
+            + "                     AND o.voided = 0 "
+            + "                     AND e.encounter_type = ${28} "
+            + "                     AND o.concept_id = ${2094} "
+            + "                     AND o.value_coded = ${703} "
+            + "                     AND e.location_id = :location "
+            + "                     AND max_suspected.patient_id = e.patient_id "
+            + "                     AND e.encounter_datetime >= "
+            + "                         max_suspected.suspected_result_date "
+            + "                     AND e.encounter_datetime BETWEEN :startDate AND :endDate)";
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java]: this verification is not needed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 10]
```diff
@@ -248,6 +818,211 @@ public CohortDefinition getPostTreatmentFollowUp(CXCASCRNResult cxcascrnResult)
...
+   *
+   * <p><b>Note:</b> If there are discrepant results registered on the same (most recent) day, the
+   * system will consider the following hierarchy:
+   *
+   * <ul>
+   *   <li>Positive
+   *   <li>Suspected Cancer
+   *   <li>Negative
+   * </ul>
+   */
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java]: replace with the correct javadocs for this method and correct the method name

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 11]
```diff
@@ -248,6 +818,211 @@ public CohortDefinition getPostTreatmentFollowUp(CXCASCRNResult cxcascrnResult)
...
+            + "                                                                e.encounter_id "
+            + "                                       WHERE  p.voided = 0 "
+            + "                                         AND e.voided = 0 "
+            + "                                         AND o.voided = 0 "
+            + "                                         AND e.encounter_datetime < :startDate "
+            + "                                         AND e.location_id = :location "
+            + "                                         AND e.encounter_type = ${28} "
+            + "                                         AND o.concept_id = ${2094} "
+            + "                                         AND o.value_coded IS NOT NULL "
+            + "                                       GROUP  BY p.patient_id) AS recent "
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java]: Confirm if you want the last result of any type or from the predicted one (RF4)

**Review by helderjosue — CHANGES_REQUESTED** [review]
Please check the recommended changes

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java (hunk 1 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java:118]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java (hunk 2 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java:141]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java (hunk 4 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java:309]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java (hunk 8 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java:330]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java (hunk 9 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java:409]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java (hunk 6 above)
Comment (TinoMuianga) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java:291]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java [hunk 12]
```diff
@@ -248,6 +798,209 @@ public CohortDefinition getPostTreatmentFollowUp(CXCASCRNResult cxcascrnResult)
...
+            + "  AND o1.voided = 0 "
+            + "  AND o2.voided = 0 "
+            + "  AND e.encounter_type = ${28} "
+            + "  AND e.location_id = :location "
+            + "  AND ( ( o1.concept_id = ${1185} "
+            + "    AND o1.value_coded IN ( ${23974}, ${165439} ) "
+            + "    AND o1.obs_datetime >= final.recent_datetime ) "
+            + "    OR ( o2.concept_id = ${2149} "
+            + "        AND o2.value_coded IN ( ${23974}, ${23972}, ${23970}, ${23973} ) "
+            + "        AND o2.obs_datetime >= final.recent_datetime ) ) ";
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/CXCASCRNCohortQueries.java]: This is not aligned with the requirement in the javadocs

**Review by helderjosue — APPROVED** [review]
LGTM

**Review by ItsMurumba — APPROVED** [review]
LGTM

**Pull request merged**